rfcs/wait_for.md

```diff
@@ -0,0 +1,60 @@
+# RFC 56: `wait_for[_callback]` methods
+
+## Summary
+
+Add `wait_for` and `wait_for_callback` methods to the wpt `Test`
+object that evaluate a condition, and resolve a promise, or call a
+callback, respectively, once the condition is true.
+
+## Details
+
+Using timers in tests is problematic; depending on the speed of the
+system the test can either wait a long time unnecessarily or be flaky
+because the operation sometimes takes longer than allowed by the
+timeout. The timeout multiplier in conjunction with `step_timeout`
+can alleviate this by adjusting the timeout to the overall system
+speed, but authors are still incentivised to provide short timeouts in
+order to allow tests to run faster.
+
+In some cases it's possible to detect directly whether the thing being
+waited on has occured or not. In this case the performance/robustness
+tradeoff can be improved by having a relatively long total timeout but
+polling for success to allow the test to proceed as soon as possible.
+
+To encourage this pattern we introduce two new methods on `Test`:
+
+`wait_for_callback(cond, callback, description, timeout=3000,
```

Comment:
Bikeshed: `wait_for_callback` sounds like it waits for a callback to be called. Not sure what would be better… `callback_wait_for`, maybe?

Reply:
I had the same thought, but also don't know what to call it instead. Maybe something involving "step" to make it clear (?) that the callback will be wrapped in a step func automatically. `t.wait_for_step`?

Reply:
Yeah, I also don't know what's really better. `callback_wait_for` perhaps, although it's a bit weird. `step` seems a bit weird because both this and the promise variant are implied steps at least when running the condition function. More verbose names would be like `step_wait_cond` `step_wait_cond_and_callback`. I guess if we dropped `_cond` it wouldn't be too bad? `step_wait` and `step_wait_and_callback`?

Reply:
Or I suppose we could assume that all modern browser devs and other test contributors should be familiar with Rust and so use `step_wait` and `step_wait_and_then` for the promise/callback versions respectively? (that's a serious proposal despite the non-serious assumption ;)